src/coreneuron/io/phase3.cpp

```diff
@@ -0,0 +1,101 @@
+/*
+# =============================================================================
+# Copyright (c) 2016 - 2024 Blue Brain Project/EPFL
+#
+# See top-level LICENSE file for details.
+# =============================================================================
+*/
+#include <algorithm>
+
+#include "coreneuron/io/phase3.hpp"
+// Where nrn2core_get_dat3_secmapping_ is declared with extern "C" to avoid symbol name mangling
+// caused by dual ABI for std::string
+#include "coreneuron/io/nrn2core_direct.h"
+
+void (*nrn2core_get_dat3_cell_count_)(int& cell_count);
+void (*nrn2core_get_dat3_cellmapping_)(int i, int& gid, int& nsec, int& nseg, int& n_seclist);
+void (*nrn2core_get_dat3_secmapping_)(int i_c,
+                                      int i_sec,
+                                      std::string& sclname,
+                                      int& nsec,
+                                      int& nseg,
+                                      size_t& total_lfp_factors,
+                                      int& n_electrodes,
+                                      std::vector<int>& data_sec,
+                                      std::vector<int>& data_seg,
+                                      std::vector<double>& data_lfp);
+
+namespace coreneuron {
+void Phase3::read_file(FileHandler& F, NrnThreadMappingInfo* ntmapping) {
+    int count = 0;
+    F.read_mapping_cell_count(&count);
+    /** for every neuron */
+    for (int i = 0; i < count; i++) {
+        int gid, nsec, nseg, nseclist;
+        // read counts
+        F.read_mapping_count(&gid, &nsec, &nseg, &nseclist);
+        CellMapping* cmap = new CellMapping(gid);
+        // read section-segment mapping for every section list
+        for (int j = 0; j < nseclist; j++) {
+            SecMapping* smap = new SecMapping();
```

Comment:
could you check if `cmap` and `smap` are freed? Same in `Phase3::read_direct()`.

Reply:
I guess that they are cleaned [here](https://github.com/neuronsimulator/nrn/blob/master/src/coreneuron/io/nrnsection_mapping.hpp#L185).